src/openms/source/ANALYSIS/OPENSWATH/SpectrumAddition.cpp

```diff
@@ -139,11 +183,78 @@ namespace OpenMS
           master_spectrum_filtered->getMZArray()->data.push_back(resampled_peak_container->getMZArray()->data[i]);
         }
       }
+
+      OPENMS_POSTCONDITION( std::adjacent_find(added_spec->getMZArray()->data.begin(),
+           added_spec->getMZArray()->data.end(), std::greater<double>()) == added_spec->getMZArray()->data.end(),
+           "Postcondition violated: m/z vector needs to be sorted!" )
+
+
       return master_spectrum_filtered;
     }
   }
 
-  OpenMS::MSSpectrum SpectrumAddition::addUpSpectra(const std::vector<OpenMS::MSSpectrum>& all_spectra, double sampling_rate, bool filter_zeros)
+
+  OpenSwath::SpectrumPtr SpectrumAddition::addUpSpectra(const SpectrumSequence& all_spectra,
+                                             const RangeMobility& im_range,
+                                             double sampling_rate,
+                                             bool filter_zeros)
+  {
+    OPENMS_PRECONDITION(! (im_range.isEmpty() || all_spectra[0].getFloatDataArrays().empty()), "Can only resample spectra with 2 data dimensions (no ion mobility spectra)")
+    OpenSwath::SpectrumPtr added_spec(new OpenSwath::Spectrum);
+
+    // If no spectra found return
+    if (all_spectra.empty())
+    {
+      return added_spec;
+    }
+
+    if (im_range.isEmpty())
+    {
+      return addUpSpectra(all_spectra, sampling_rate, filter_zeros);
+    }
+    // since resampling is not supported on 3D data first filter by drift time (if possible) and then add
+    // (!im_range.isEmpty())
+    SpectrumSequence filteredSpectra;
+    for (auto spec: all_spectra)
+    {
+      filteredSpectra.push_back(OpenSwath::ISpectrumAccess::filterByDrift(spec, im_range.getMin(), im_range.getMax()));
+    }
+    return addUpSpectra(filteredSpectra, sampling_rate, filter_zeros);
+  }
+
+  OpenSwath::SpectrumPtr SpectrumAddition::concatenateSpectra(const SpectrumSequence& all_spectra)
+
+  {
+    OpenSwath::SpectrumPtr added_spec(new OpenSwath::Spectrum);
+    // Ensure that we have the same number of data arrays as in the input spectrum
+    // copying the extra data arrays descriptions onto the added spectra
+    if (!all_spectra.empty() && all_spectra[0]->getDataArrays().size() > 2)
+    {
+      for (Size k = 2; k < all_spectra[0]->getDataArrays().size(); k++)
+      {
+        OpenSwath::BinaryDataArrayPtr tmp (new OpenSwath::BinaryDataArray());
+        tmp->description = all_spectra[0]->getDataArrays()[k]->description;
+        added_spec->getDataArrays().push_back(tmp);
+      }
+    }
+
+    // Simply concatenate all spectra together and sort in the end
+    for (const auto& s : all_spectra)
+    {
+      for (Size k = 0; k < s->getDataArrays().size(); k++)
+      {
+        auto& v1 = added_spec->getDataArrays()[k]->data;
+        auto& v2 = s->getDataArrays()[k]->data;
+
+        v1.reserve( v1.size() + v2.size() );
+        v1.insert( v1.end(), v2.begin(), v2.end() );
+      }
+    }
+    sortSpectrumByMZ(*added_spec);
```

Comment:
I guess that all spectra are already sorted by m/z? If yes, you can make this much faster by simply doing an inplace_merge.

Reply:
Not sure how to do an inplace is feasible considering that the  intensity as these will not be sorted. code might get quite messy...

Reply:
not sure what you mean.
Why would unsorted intensity be a problem when you trying to sort by m/z?

Reply:
I am not sure what the best strategy would be to preserve the (mz, intensity) pairings with an inplace merge

Reply:
can an in place merge be done without screwing up the mz-intensity pairing considering that mz and intensity are stored in independent arrays

Reply:
see [MSSpectrum::sortByPositionPresorted(...)](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1MSSpectrum.html#a943f07f7e779e5a27bc6bc5f70b87942),  for an example.
It depends on how much time the sort actually takes and if its worth the trouble to port it to the swath spectrum.

Reply:
Thank you for the example, that makes sense. From my experience the current method is not used too often in practice and is more experimental so I think that the current method is sufficient for the time being.